File: server/src/controllers/issueController.js
const Issue = require("../models/Issue");
const Project = require("../models/Project");
const User = require("../models/User");


// =====================================================
// CREATE ISSUE
// =====================================================

const createIssue = async (req, res) => {
    try {
        console.log("=================================");
        console.log("CREATE ISSUE CONTROLLER CALLED");
        console.log("Request body:", req.body);

        const {
            title,
            description,
            type,
            priority,
            projectId,
            assignee,
            dueDate,
        } = req.body;

        // 1. Validate required fields
        if (!title || !projectId) {
            return res.status(400).json({
                success: false,
                message: "Title and projectId are required",
            });
        }

        // 2. Check project
        const project = await Project.findById(projectId);

        if (!project) {
            return res.status(404).json({
                success: false,
                message: "Project not found",
            });
        }

        // 3. Check assignee if provided
        if (assignee) {
            const user = await User.findById(assignee);

            if (!user) {
                return res.status(404).json({
                    success: false,
                    message: "Assignee not found",
                });
            }
        }

        // 4. Create issue
        const issue = await Issue.create({
            title: title.trim(),
            description: description ? description.trim() : "",
            type: type || "TASK",
            priority: priority || "MEDIUM",
            status: "TODO",
            projectId,
            assignee: assignee || null,
            dueDate: dueDate || null,
        });

        console.log("Issue created:", issue._id);

        // 5. Fetch populated issue
        const populatedIssue = await Issue.findById(issue._id)
            .populate("projectId", "name key")
            .populate("assignee", "name email");

        return res.status(201).json({
            success: true,
            message: "Issue created successfully",
            data: {
                issue: populatedIssue,
            },
        });

    } catch (error) {
        console.error("Create issue error:", error);

        return res.status(500).json({
            success: false,
            message: "Server error while creating issue",
            error: error.message,
        });
    }
};

const getAllIssues = async (req, res) => {
    try {
        console.log("GET ALL ISSUES CONTROLLER CALLED");

        const issues = await Issue.find()
            .populate("projectId", "name key")
            .populate("assignee", "name email")
            .sort({ createdAt: -1 });

        return res.status(200).json({
            success: true,
            count: issues.length,
            data: {
                issues,
            },
        });

    } catch (error) {
        console.error("Get all issues error:", error);

        return res.status(500).json({
            success: false,
            message: "Server error while fetching issues",
            error: error.message,
        });
    }
};

// GET ISSUES BY PROJECT

const getIssuesByProject = async (req, res) => {
    try {
        console.log("GET ISSUES BY PROJECT CONTROLLER CALLED");

        const { projectId } = req.params;

        if (!projectId) {
            return res.status(400).json({
                success: false,
                message: "Project ID is required",
            });
        }

        const project = await Project.findById(projectId);

        if (!project) {
            return res.status(404).json({
                success: false,
                message: "Project not found",
            });
        }

        const issues = await Issue.find({
            projectId,
        })
            .populate("projectId", "name key")
            .populate("assignee", "name email")
            .sort({ createdAt: -1 });

        return res.status(200).json({
            success: true,
            count: issues.length,
            data: {
                issues,
            },
        });

    } catch (error) {
        console.error("Get project issues error:", error);

        return res.status(500).json({
            success: false,
            message: "Server error while fetching project issues",
            error: error.message,
        });
    }
};

// UPDATE ISSUE
// =====================================================

const updateIssue = async (req, res) => {
    try {
        console.log("UPDATE ISSUE CONTROLLER CALLED");
        console.log("Issue ID:", req.params.id);
        console.log("Request body:", req.body);

        const { id } = req.params;

        const {
            title,
            description,
            type,
            priority,
            status,
            assignee,
            dueDate,
        } = req.body;

        // 1. Find issue
        const issue = await Issue.findById(id);

        if (!issue) {
            return res.status(404).json({
                success: false,
                message: "Issue not found",
            });
        }

        // 2. Update only provided fields
        if (title !== undefined) {
            issue.title = title.trim();
        }

        if (description !== undefined) {
            issue.description = description;
        }

        if (type !== undefined) {
            issue.type = type;
        }

        if (priority !== undefined) {
            issue.priority = priority;
        }

        if (status !== undefined) {
            issue.status = status;
        }

        if (assignee !== undefined) {
            if (assignee) {
                const user = await User.findById(assignee);

                if (!user) {
                    return res.status(404).json({
                        success: false,
                        message: "Assignee not found",
                    });
                }
            }

            issue.assignee = assignee || null;
        }

        if (dueDate !== undefined) {
            issue.dueDate = dueDate || null;
        }

        // 3. Save changes
        await issue.save();

        // 4. Populate related data
        const updatedIssue = await Issue.findById(issue._id)
            .populate("projectId", "name key")
            .populate("assignee", "name email");

        return res.status(200).json({
            success: true,
            message: "Issue updated successfully",
            data: {
                issue: updatedIssue,
            },
        });

    } catch (error) {
        console.error("Update issue error:", error);

        return res.status(500).json({
            success: false,
            message: "Server error while updating issue",
            error: error.message,
        });
    }
};

module.exports = {
    createIssue,
    getAllIssues,
    getIssuesByProject,
    updateIssue,
};

File: server/src/main/java/com/example/jira/controller/IssueController.java
package com.example.jira.controller;

import com.example.jira.dto.ApiResponse;
import com.example.jira.dto.CreateIssueRequest;
import com.example.jira.dto.UpdateIssueRequest;
import com.example.jira.repository.ProjectRepository;
import com.example.jira.repository.UserRepository;
import org.bson.Document;
import org.bson.types.ObjectId;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.*;

@RestController
@RequestMapping("/api/issues")
public class IssueController {

    @Autowired
    private ProjectRepository projectRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private MongoTemplate mongoTemplate;

    @GetMapping
    public ResponseEntity<ApiResponse> getAllIssues() {
        try {
            List<Map<String, Object>> resultList = new ArrayList<>();
            List<Document> documents = mongoTemplate.findAll(Document.class, "issues");

            for (Document doc : documents) {
                Map<String, Object> map = convertDocumentToMap(doc);
                resultList.add(map);
            }

            Map<String, Object> data = new HashMap<>();
            data.put("issues", resultList);
            return ResponseEntity.ok(ApiResponse.successWithCount(resultList.size(), data));
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Server error while fetching issues", e.getMessage()));
        }
    }

    @PostMapping
    public ResponseEntity<ApiResponse> createIssue(@RequestBody CreateIssueRequest request) {
        try {
            if (request.getTitle() == null || request.getTitle().trim().isEmpty() ||
                request.getProjectId() == null || request.getProjectId().trim().isEmpty()) {
                return ResponseEntity.badRequest()
                        .body(ApiResponse.error("Title and projectId are required"));
            }

            Document doc = new Document();
            doc.put("title", request.getTitle().trim());
            doc.put("description", request.getDescription() != null ? request.getDescription().trim() : "");
            doc.put("type", request.getType() != null ? request.getType() : "TASK");
            doc.put("priority", request.getPriority() != null ? request.getPriority() : "MEDIUM");
            doc.put("status", "TODO");

            // Attach project reference
            String pId = request.getProjectId();
            projectRepository.findById(pId).ifPresentOrElse(
                p -> {
                    Map<String, Object> pMap = new HashMap<>();
                    pMap.put("_id", p.getId());
                    pMap.put("name", p.getName());
                    pMap.put("key", p.getKey());
                    doc.put("projectId", pMap);
                },
                () -> doc.put("projectId", pId)
            );

            // Attach assignee reference
            if (request.getAssignee() != null && !request.getAssignee().trim().isEmpty()) {
                String uId = request.getAssignee().trim();
                userRepository.findById(uId).ifPresentOrElse(
                    u -> {
                        Map<String, Object> uMap = new HashMap<>();
                        uMap.put("_id", u.getId());
                        uMap.put("name", u.getName());
                        uMap.put("email", u.getEmail());
                        doc.put("assignee", uMap);
                    },
                    () -> doc.put("assignee", uId)
                );
            }

            doc.put("dueDate", request.getDueDate());
            doc.put("createdAt", new Date());
            doc.put("updatedAt", new Date());

            mongoTemplate.save(doc, "issues");

            Map<String, Object> map = convertDocumentToMap(doc);
            Map<String, Object> data = new HashMap<>();
            data.put("issue", map);

            return ResponseEntity.status(HttpStatus.CREATED)
                    .body(ApiResponse.success("Issue created successfully", data));

        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Server error while creating issue", e.getMessage()));
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse> updateIssue(@PathVariable String id, @RequestBody UpdateIssueRequest request) {
        try {
            Document doc = null;
            if (ObjectId.isValid(id)) {
                doc = mongoTemplate.findById(new ObjectId(id), Document.class, "issues");
            }
            if (doc == null) {
                doc = mongoTemplate.findById(id, Document.class, "issues");
            }

            if (doc == null) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(ApiResponse.error("Issue not found"));
            }

            if (request.getTitle() != null) {
                doc.put("title", request.getTitle().trim());
            }
            if (request.getDescription() != null) {
                doc.put("description", request.getDescription().trim());
            }
            if (request.getType() != null) {
                doc.put("type", request.getType());
            }
            if (request.getPriority() != null) {
                doc.put("priority", request.getPriority());
            }
            if (request.getStatus() != null) {
                doc.put("status", request.getStatus());
            }
            if (request.getDueDate() != null) {
                doc.put("dueDate", request.getDueDate());
            }
            if (request.getAssignee() != null) {
                if (!request.getAssignee().trim().isEmpty()) {
                    String uId = request.getAssignee().trim();
                    userRepository.findById(uId).ifPresentOrElse(
                        u -> {
                            Map<String, Object> uMap = new HashMap<>();
                            uMap.put("_id", u.getId());
                            uMap.put("name", u.getName());
                            uMap.put("email", u.getEmail());
                            doc.put("assignee", uMap);
                        },
                        () -> doc.put("assignee", uId)
                    );
                } else {
                    doc.put("assignee", null);
                }
            }

            doc.put("updatedAt", new Date());
            mongoTemplate.save(doc, "issues");

            Map<String, Object> map = convertDocumentToMap(doc);
            Map<String, Object> data = new HashMap<>();
            data.put("issue", map);

            return ResponseEntity.ok(ApiResponse.success("Issue updated successfully", data));

        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Server error while updating issue", e.getMessage()));
        }
    }

    @GetMapping("/project/{projectId}")
    public ResponseEntity<ApiResponse> getIssuesByProject(@PathVariable String projectId) {
        try {
            List<Map<String, Object>> resultList = new ArrayList<>();
            List<Document> documents = mongoTemplate.findAll(Document.class, "issues");

            for (Document doc : documents) {
                Object pIdObj = doc.get("projectId");
                boolean matches = false;
                if (pIdObj instanceof Document) {
                    Object nestedId = ((Document) pIdObj).get("_id");
                    if (nestedId != null && nestedId.toString().equals(projectId)) {
                        matches = true;
                    }
                } else if (pIdObj != null && pIdObj.toString().equals(projectId)) {
                    matches = true;
                }

                if (matches) {
                    resultList.add(convertDocumentToMap(doc));
                }
            }

            Map<String, Object> data = new HashMap<>();
            data.put("issues", resultList);
            return ResponseEntity.ok(ApiResponse.successWithCount(resultList.size(), data));
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Server error while fetching project issues", e.getMessage()));
        }
    }

    private Map<String, Object> convertDocumentToMap(Document doc) {
        Map<String, Object> map = new HashMap<>(doc);
        Object idObj = map.get("_id");
        if (idObj != null) {
            map.put("_id", idObj.toString());
            map.put("id", idObj.toString());
        }
        return map;
    }
}
